File: Java/src/_0802_Find_Eventual_Safe_States/Solution.java
package _0802_Find_Eventual_Safe_States;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

/**
 * Date: 2021/8/5 11:03
 * Content:
 */
public class Solution {

    public List<Integer> eventualSafeNodes(int[][] graph) {
        int n = graph.length;
        int[] d = new int[n];
        List<List<Integer>> g = new ArrayList<>();
        for (int i = 0; i < n; i++) g.add(new ArrayList<>());
        for (int i = 0; i < n; i++)
            for (int b : graph[i]) {
                int a = i;
                g.get(b).add(a);
                d[a]++;  // a的入度增加一
            }

        Queue<Integer> q = new LinkedList<>();
        for (int i = 0; i < n; i++)
            if (d[i] == 0)
                q.add(i);
        while (!q.isEmpty()) {
            int t = q.remove();
            if (g.get(t).isEmpty()) continue;
            for (int i : g.get(t))
                if (--d[i] == 0)
                    q.add(i);
        }

        List<Integer> res = new ArrayList<>();
        for (int i = 0; i < n; i++)
            if (d[i] == 0)
                res.add(i);
        return res;
    }

    public static void main(String[] args) {

        int[][] graph = {
                {1, 2},
                {2, 3},
                {5},
                {0},
                {5},
                {},
                {},
        };
        System.out.println((new Solution()).eventualSafeNodes(graph));  // [2, 4, 5, 6]
    }
}

File: Python/_0802_Find_Eventual_Safe_States/Solution.py
# coding=utf-8
# Date: 2021/8/5 11:12
from typing import List


# 执行用时：156 ms, 在所有 Python3 提交中击败了61.27%的用户
# 内存消耗：19.9 MB, 在所有 Python3 提交中击败了40.46%的用户
class Solution:
    def eventualSafeNodes(self, graph: List[List[int]]) -> List[int]:
        n = len(graph)
        d = [0] * n
        g = [[] for _ in range(n)]
        for i in range(n):
            for b in graph[i]:
                a = i
                g[b].append(a)
                d[a] += 1

        q = []
        for i in range(n):
            if d[i] == 0:
                q.append(i)

        while len(q) != 0:
            t = q[0]
            q.pop(0)
            for i in g[t]:
                d[i] -= 1
                if d[i] == 0:
                    q.append(i)

        res = []
        for i in range(n):
            if d[i] == 0:
                res.append(i)
        return res


if __name__ == "__main__":
    print(Solution().eventualSafeNodes([[], [0, 2, 3, 4], [3], [4], []]))
